src/serac/physics/materials/parameterized_solid_material.hpp

```diff
@@ -105,4 +105,95 @@ struct ParameterizedNeoHookeanSolid {
   double G0;       ///< base shear modulus
 };
 
+/**
+ * @brief Infers type resulting from algebraic expressions of a group of variables
+ *
+ * Useful if one needs to create a variable that is dual-valued if any operands are dual.
+ */
+template <typename... T>
+struct underlying_scalar {
+  using type = decltype((T{} + ...));  ///< type of the sum of the parameters
+};
+
+/// @brief J2 material with Voce hardening, with hardening parameters exposed as differentiable parameters
+struct ParameterizedJ2Nonlinear {
+  static constexpr int    dim = 3;      ///< dimension of space
+  static constexpr double tol = 1e-10;  ///< relative tolerance on residual for accepting return map solution
+
+  double E;        ///< Young's modulus
+  double nu;       ///< Poisson's ratio
+  double density;  ///< mass density
+
+  /// @brief variables required to characterize the hysteresis response
+  struct State {
+    tensor<double, dim, dim> plastic_strain;              ///< plastic strain
+    double                   accumulated_plastic_strain;  ///< uniaxial equivalent plastic strain
+  };
+
+  /// @brief calculate the Cauchy stress, given the displacement gradient and previous material state
+  template <typename DisplacementGradient, typename YieldStrength, typename SaturationStrength, typename StrainConstant>
+  auto operator()(State& state, const DisplacementGradient du_dX, const YieldStrength sigma_y,
+                  const SaturationStrength sigma_sat, const StrainConstant strain_constant) const
+  {
+    // The output stress tensor should use dual numbers if any of the parameters are dual.
+    // This slightly ugly trick to accomplishes that by picking up any dual number types
+    // from the parameters into the dummy variable "one".
+    // Another possiblity would be to cast the results to the correct type at the end, which
+    // would avoid doing any unneccessary arithmetic with dual numbers.
+    using T = typename underlying_scalar<YieldStrength, SaturationStrength, StrainConstant>::type;
+    T one;
+    one = 1.0;
```

Comment:
first, this ugly workaround was my suggestion, so I apologize.

>    // Another possiblity would be to cast the results to the correct type at the end, which
>    // would avoid doing any unneccessary arithmetic with dual numbers.

Two ways I can think of to do this are mocked up here: https://godbolt.org/z/v4hK9EoTP

-Option 0 (implicit conversions) make the written code a little bit cleaner to look at, but have the usual drawbacks of implicit conversions (potentially unintended type changes). That being said, this kind of implicit conversion is the least problematic-- analogous to how `float f = 3.0;` is technically an implicit conversion, but is harmless in practice. This approach also has the downside of requiring us to reimplement the compiler-generated constructors for the tensor class.

-Option 1 (explicit conversions) make the written code a little bit more verbose, but also makes it clearer about where potential type changes are occurring.

Reply:
I think I'm leaning toward option 0, but I'm fine either way. I think this should happen outside of this PR though.